File: COMP-3270/main.py
import math
import random
import time
import matplotlib.pyplot as plt
from typing import List


names = ["algorithm-1", "algorithm-2", "algorithm-3", "algorithm-4"]


def algorithm_1(array: List[int]):
    max_so_far = 0
    for L in range(len(array)):
        for U in range(L, len(array)):
            number_sum = 0
            for i in range(L, U):
                number_sum = number_sum + array[i]
            max_so_far = max(max_so_far, number_sum)
    return max_so_far


def algorithm_2(array: List[int]):
    max_so_far = 0
    for L in range(len(array)):
        number_sum = 0
        for U in range(L, len(array)):
            number_sum = number_sum + array[U]
            max_so_far = max(max_so_far, number_sum)
    return max_so_far


def max_sum(array: List[int], k, u: int):
    if k > u:
        return 0
    if k == u:
        return max(0, array[k])
    m = math.floor((k + u) / 2)
    number_sum = 0
    max_to_left = 0
    for i in range(m, k, -1):
        number_sum = number_sum + array[i]
        max_to_left = max(max_to_left, number_sum)

    number_sum = 0
    max_to_right = 0
    for i in range(m+1, u):
        number_sum = number_sum + array[i]
        max_to_right = max(max_to_right, number_sum)

    max_crossing = max_to_left + max_to_right
    max_in_a = max_sum(array, k, m)
    max_in_b = max_sum(array, m + 1, u)
    return max(max_crossing, max_in_a, max_in_b)


def algorithm_4(array: List[int]):
    max_so_far = 0
    max_ending_here = 0
    for i in range(len(array)):
        max_ending_here = max(0, max_ending_here + array[i])
        max_so_far = max(max_so_far, max_ending_here)
    return max_so_far


def read_input_file():
    """ Reads in the input file and then calls console_out to then print the results from
    running the file input in the above algorithms
    """
    numbers = []
    with open('phw_input.txt', 'r') as reader:
        strings = reader.read().split(",")
        for i in range(len(strings)):
            numbers.append(int(strings[i]))
    console_out(numbers)


def console_out(numbers: List[int]):
    """ Takes in a list of ints and then will print the results of the list when run through the above algos"""
    results = []
    results.append(algorithm_1(numbers))
    results.append(algorithm_2(numbers))
    results.append(max_sum(numbers, 0, len(numbers)-1))
    results.append(algorithm_4(numbers))

    for i in range(len(names)):
        print(names[i] + ": " + str(results[i]) + ";", end=" ")


def driver():
    """ Driver program
    Section 1: Completes the first requirement of the project by taking in a
    input file and then printing those results to the console.

    Section 2: Creation of helper list and variables

    Section 3: Iterates through the list size 10-100, some predetermined about of times,
    and stores those results in the variable results

    Section 4: Prints the indices from results to the console and to a file.

    Section 5: Creates the graphs
    """
    # Section 1
    read_input_file()

    # Section 2
    container = []
    for i in range(19):
        temp = []
        for j in range(10 + (i*5)):
            temp.append(random.randint(-100, 100))
        container.append(temp)

    results = [[], [], [], [], [], [], [], []]
    print("")
    runs = 300
    divisor_run = 1000000000
    divisor_theory = 1000000
    # Section 3
    time_start_total = time.time_ns()
    for i in range(len(container)):
        time_start = time.time_ns()
        for j in range(runs):
            algorithm_1(container[i])
        time_end = time.time_ns()
        results[0].append((time_end - time_start) / divisor_run)
        n = len(container[i])
        results[4].append((3*(n**3) + 5*(n**2) + 6*n + 4) / divisor_theory)
    print("Made it past the first")
    for i in range(len(container)):
        time_start = time.time_ns()
        for j in range(runs):
            algorithm_2(container[i])
        time_end = time.time_ns()
        results[1].append((time_end - time_start) / divisor_run)
        n = len(container[i])
        results[5].append(1.5*(15*(n**2) + 8*n + 16) / divisor_theory)
    print("Past the second")
    for i in range(len(container)):
        time_start = time.time_ns()
        for j in range(runs):
            max_sum(container[i], 0, (len(container[i])-1))
        time_end = time.time_ns()
        results[2].append((time_end - time_start) / divisor_run)
        n = len(container[i])
        results[6].append((2*(n/2)+14*n+42) / 10000)
    print("Made it past the third")
    for i in range(len(container)):
        time_start = time.time_ns()
        for j in range(runs):
            algorithm_4(container[i])
        time_end = time.time_ns()
        results[3].append((time_end - time_start) / divisor_run)
        n = len(container[i])
        results[7].append((17*n + 12) / 100000)
    time_end_total = time.time_ns()

    # Section 4
    for i in range(len(results[0])):
        print(results[0][i], results[1][i], results[2][i], results[3][i],
              results[4][i], results[5][i], results[6][i], results[7][i])

    with open('perrybunn_phw_output.txt', 'w') as file:
        for i in range(len(results[0])):
            file.write(f'{results[0][i]}, {results[1][i]}, {results[2][i]}, {results[3][i]}, ' +
                       f'{results[4][i]}, {results[5][i]}, {results[6][i]}, {results[7][i]}\n')
        file.close()

    # Section 5
    run_size = []
    for i in range(19):
        run_size.append(5*(2+i))

    for i in range(len(results)-4):
        plt.plot(run_size, results[i], label='Experimental')
        plt.plot(run_size, results[i+4], label='Theoretical')
        plt.title(names[i])
        plt.ylabel('seconds')
        plt.xlabel('n')
        plt.legend(loc='upper left')
        plt.savefig(f'{names[i]}-graph.png')
        plt.show()

    number_sum = 0
    for i in range(len(results)-4):
        for j in range(len(results[i])):
            number_sum += results[i][j]
    print((time_end_total - time_start_total) / 1000000000, number_sum)


if __name__ == '__main__':
    driver()

File: COMP-3270/main.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <chrono>
#include <cmath>

using namespace std;

int Algorithm1(int X[]){
    int maxSoFar = 0;
    for(int i = 0; i < sizeof(X); i++){
        for(int j = i; j < sizeof(X);j++){
            int sum = 0;
            for(int k = i; k < j; k++){
                sum = sum + X[k];
            }
            maxSoFar = max(maxSoFar,sum);
        }
    }
    return maxSoFar;
}

int Algorithm2(int X[]){
    int maxSoFar = 0;
    for(int i = 0; i < sizeof(X); i++){
        int sum = 0;
        for(int j = i; j < sizeof(X); j++){
            sum = sum + X[j];
            maxSoFar = max(maxSoFar,sum);
        }
    }
    return maxSoFar;
}

int Algorithm3(int X[],int L, int U){
    if(L > U){
        return 0;
    }
    if(L==U){
        return max(0,X[L]);
    }
    int M = (L+U)/2;
    int sum = 0;
    int maxToLeft = 0;
    for(int I = M; I > L; I--){
        sum = sum+ X[I];
        maxToLeft = max(maxToLeft,sum);
    }
    sum = 0;
    int maxToRight = 0;
    for (int I = M+1; I < U; I++){
        sum = sum+X[I];
        maxToRight = max(maxToRight,sum);
    }
    int maxCrossing = maxToLeft + maxToRight;
    int maxInA = Algorithm3(X,L,M);
    int maxInB = Algorithm3(X,M+1,U);
    int maxOfAB = max(maxInA,maxInB);
    return max(maxCrossing, maxOfAB);
}

int Algorithm4(int X[]){
    int maxSoFar = 0;
    int maxEndingHere = 0;
    for(int i = 0; i < sizeof(X); i++){
        maxEndingHere = max(0,maxEndingHere+X[i]);
        maxSoFar = max(maxSoFar, maxEndingHere);
    }
    return maxSoFar;
}

int main() {
//    First, the main program should read from a file named “phw_input.txt” containing 10 comma-delimited integers in the first line,create an array containing these 10 integers, and run each of the algorithms on that input array, and print out the answer produced by each on the console as follows:"algorithm-1: <answer>; algorithm-2:<answer>;algorithm-3:<answer>; algorithm-4:<answer> where <answer> is the MSCS as determined by each of the algorithms.

    int tenIntegers[10];
    fstream file;
    file.open("phw_input.txt", ifstream::in);
    string temp;
    for(int i = 0; i < 10; i++ ){
        getline(file, temp,',');
        tenIntegers[i] = (int)stoi(temp);
    }
    file.close();

//    call algos here

    cout << "algorithm-1: " << Algorithm1(tenIntegers) << ", ";
    cout << "algorithm-2: " << Algorithm2(tenIntegers) << ", ";
    cout << "algorithm-3: " << Algorithm3(tenIntegers,0,9) << ", ";
    cout << "algorithm-4: " << Algorithm4(tenIntegers) << "\n";

//    Next,create 19 integer sequences of length10,15,20,25,......90,95,100, containing randomly generated negative, zero and positive integers, and store these in 19arrays of size 10,15,...,95,100

    int array10[10];
    for (int i = 0; i < 10; ++i) {
        array10[i] = rand() % 100 + (-50);
    }
    int array15[15];
    for (int i = 0; i < 15; ++i) {
        array15[i] = rand() % 100 + (-50);
    }
    int array20[20];
    for (int i = 0; i < 20; ++i) {
        array20[i] = rand() % 100 + (-50);
    }
    int array25[25];
    for (int i = 0; i < 25; ++i) {
        array25[i] = rand() % 100 + (-50);
    }
    int array30[30];
    for (int i = 0; i < 30; ++i) {
        array30[i] = rand() % 100 + (-50);
    }
    int array35[35];
    for (int i = 0; i < 35; ++i) {
        array35[i] = rand() % 100 + (-50);
    }
    int array40[40];
    for (int i = 0; i < 40; ++i) {
        array40[i] = rand() % 100 + (-50);
    }
    int array45[45];
    for (int i = 0; i < 45; ++i) {
        array45[i] = rand() % 100 + (-50);
    }
    int array50[50];
    for (int i = 0; i < 50; ++i) {
        array50[i] = rand() % 100 + (-50);
    }
    int array55[55];
    for (int i = 0; i < 55; ++i) {
        array55[i] = rand() % 100 + (-50);
    }
    int array60[60];
    for (int i = 0; i < 60; ++i) {
        array60[i] = rand() % 100 + (-50);
    }
    int array65[65];
    for (int i = 0; i < 65; ++i) {
        array65[i] = rand() % 100 + (-50);
    }
    int array70[70];
    for (int i = 0; i < 70; ++i) {
        array70[i] = rand() % 100 + (-50);
    }
    int array75[75];
    for (int i = 0; i < 75; ++i) {
        array75[i] = rand() % 100 + (-50);
    }
    int array80[80];
    for (int i = 0; i < 80; ++i) {
        array80[i] = rand() % 100 + (-50);
    }
    int array85[85];
    for (int i = 0; i < 85; ++i) {
        array85[i] = rand() % 100 + (-50);
    }
    int array90[90];
    for (int i = 0; i < 90; ++i) {
        array90[i] = rand() % 100 + (-50);
    }
    int array95[95];
    for (int i = 0; i < 95; ++i) {
        array95[i] = rand() % 100 + (-50);
    }
    int array100[100];
    for (int i = 0; i < 100; ++i) {
        array100[i] = rand() % 100 + (-50);
    }

     int* container[19] = {array10,
                       array15,
                       array20,
                       array25,
                       array30,
                       array35,
                       array40,
                       array45,
                       array50,
                       array55,
                       array60,
                       array65,
                       array70,
                       array75,
                       array80,
                       array85,
                       array90,
                       array95,
                       array100};

//    Then use the system clock to measure time t1, run one of the four algorithms on array Ai(starting with i=1)N times (where N is at least100, but if your system clock does not have a good resolution you may need N to be larger, like 500 or 1000in order to get meaningful running times), then measure time t2, and compute average time needed by that algorithm to solve the problem with input size = size of Ai. Do this for each of the algorithms executing on each of the 19input arrays to fill the first four columns of a19X8matrix of integers with average execution times.Each row of this matrix corresponds to one input size, from 10-100.

    // 1-19:0-3 real time of above arrays

    double times[19][8];

    for (int j = 0; j < 19; ++j) {
        double timeAVG = 0;
        for (int i = 0; i < 10000; ++i) {
            auto start = chrono::high_resolution_clock::now();
            Algorithm1(container[j]);
            auto finish = chrono::high_resolution_clock::now();
            timeAVG += chrono::duration_cast<chrono::nanoseconds>(finish-start).count();
        }
        times[j][0] = timeAVG/10000;
        times[j][4] = (double)ceil(pow(sizeof(container[j]),3)*10000);
    }

    for (int j = 0; j < 19; ++j) {
        double timeAVG = 0;
        for (int i = 0; i < 10000; ++i) {
            auto start = chrono::high_resolution_clock::now();
            Algorithm2(container[j]);
            auto finish = chrono::high_resolution_clock::now();
            timeAVG += chrono::duration_cast<chrono::nanoseconds>(finish-start).count();
        }
        times[j][1] = timeAVG/10000;
        times[j][5] = (double)ceil(pow(sizeof(container[j]),2)*10000);
    }

    for (int j = 0; j < 19; ++j) {
        double timeAVG = 0;
        for (int i = 0; i < 10000; ++i) {
            auto start = chrono::high_resolution_clock::now();
            Algorithm3(container[j],0,sizeof(container[j])-1);
            auto finish = chrono::high_resolution_clock::now();
            timeAVG += chrono::duration_cast<chrono::nanoseconds>(finish-start).count();
        }
//        ceil(length * log(length) / log(2))) * 1000;
        times[j][2] = timeAVG/10000;
        times[j][6] = (double)ceil(sizeof(container[j]) * log(sizeof(container[j])) / log(2)) * 10000;
    }

    for (int j = 0; j < 19; ++j) {
        double timeAVG = 0;
        for (int i = 0; i < 10000; ++i) {
            auto start = chrono::high_resolution_clock::now();
            Algorithm4(container[j]);
            auto finish = chrono::high_resolution_clock::now();
            timeAVG += chrono::duration_cast<chrono::nanoseconds>(finish-start).count();
        }
        times[j][3] = timeAVG/10000;
        times[j][7] = (double)ceil(sizeof(container[j])*10000);
    }

//    Fill the last four columns of this matrix with values ceiling(T1(n)), ceiling(T2(n)), ceiling(T3(n)),and ceiling(T4(n)) where n=each input size and T(n) are the polynomials representing the theoretically calculated complexity of the three algorithms that you determined in step 2part (b). So,column 1 will have measured running times of your first algorithm and column 5will have the calculated complexity for the same algorithm; similarly for columns 2 & 6, 3 & 7, and 4&8.You may need to scale the complexity values (or use an appropriate time unit such as nano/micro/milli seconds for the measured running times) in order to bring all data into similar ranges.

    // 1-19:4-7 theoretical times

//    Your main program should write one text line of algorithm and complexity order titles separated by commas (e.g., "algorithm-1,algorithm-2,algorithm-3,algorithm-4,T1(n),T2(n),T3(n),T4(n)"), followed by the above matrix also in comma-delimited format (19lines of 8integers separated by commas) to a file called "yourname_phw_output.txt".

    fstream fileOut;
    fileOut.open("perrybunn_phw_output.txt", ifstream::out);

    fileOut << "algorithm-1, algorithm-2, algorithm-3, algorithm-4, T1(n), T2(n), T3(n), T4(n)\n";
    cout << "algorithm-1, algorithm-2, algorithm-3, algorithm-4, T1(n), T2(n), T3(n), T4(n)\n";

    for (int i = 0; i < 19; ++i) {
        for (int j = 0; j < 8; ++j) {
            if (j!=7){
                fileOut << times[i][j] << ", ";
                cout << times[i][j] << ", ";
            } else {
                fileOut << times[i][j] << "\n";
                cout << times[i][j] << "\n";
            }
        }
    }

    fileOut.close();

    return 0;
}
